Add AutoPackets documentation and fix navigation issue hiding Multiplayer Gem docs

AMZN-puvvadar opened the pull request:
This adds documentation around AutoPackets (we have docs for AutoComponents but not Packets) and fixes an issue I observed where the Multiplayer Gem doesn't appear in the left hand navigation bar. The content is searchable but can't be obviously navigated to.

lmbr-pip commented on content/docs/user-guide/gems/reference/multiplayer/_index.md:
```diff
@@ -14,10 +14,11 @@ For a quick introduction of the O3DE Multiplayer Gem and networking, watch the v
-| [Multiplayer Autocomponents](./autocomponents) | Automatically create components for use with the Multiplayer Gem using the AzAutoGen system. |
-| [Test Multiplayer Projects in the Editor](./test-in-editor) | Automatically launch local servers or connect to a remote persistent server when working on a multiplayer project in the O3DE Editor. |
+| [Overview](./multiplayer-gem/overview) | An overview of the O3DE Multiplayer Gem, how it works, and how to use it in your project. |
+| [Project configuration](./multiplayer-gem/configuration) | How to enable the O3DE Multiplayer Gem in a project. |
```
nit: Drop the O3DE part of "O3DE Multiplayer Gem" here as you drop it for all later lines.

lmbr-pip commented on content/docs/user-guide/gems/reference/multiplayer/_index.md:
```diff
@@ -14,10 +14,11 @@ For a quick introduction of the O3DE Multiplayer Gem and networking, watch the v
+| [Overview](./multiplayer-gem/overview) | An overview of the O3DE Multiplayer Gem, how it works, and how to use it in your project. |
+| [Project configuration](./multiplayer-gem/configuration) | How to enable the O3DE Multiplayer Gem in a project. |
+| [Multiplayer AutoComponents](./multiplayer-gem/autocomponents) | Automatically create components for use with the Multiplayer Gem using the AzAutoGen system. |
+| [Multiplayer AutoPackets](./multiplayer-gem/autopackets) | Define and automatically create packets for use with the Multiplayer Gem using the AzAutoGen system. |
```
Is there a descriptive word to place in front of "packets". From this sentence its unclear why I need to do/should do this.  But what is a "packet" in this context?

AMZN-puvvadar replied:
I'm moving this to AzNetworking where it'll make a lot more sense.

lmbr-pip commented on content/docs/user-guide/gems/reference/multiplayer/_index.md:
```diff
@@ -14,10 +14,11 @@ For a quick introduction of the O3DE Multiplayer Gem and networking, watch the v
-| [Test Multiplayer Projects in the Editor](./test-in-editor) | Automatically launch local servers or connect to a remote persistent server when working on a multiplayer project in the O3DE Editor. |
+| [Overview](./multiplayer-gem/overview) | An overview of the O3DE Multiplayer Gem, how it works, and how to use it in your project. |
+| [Project configuration](./multiplayer-gem/configuration) | How to enable the O3DE Multiplayer Gem in a project. |
+| [Multiplayer AutoComponents](./multiplayer-gem/autocomponents) | Automatically create components for use with the Multiplayer Gem using the AzAutoGen system. |
```
IMHO these are technical links, but they do not provide any context for why a developer wants to do this thing/

What is a "component" here? Firstly, component is standard terminology  from Lumberyard/O3DE days, but this is really an Auto component. Hard to tell from sentence the why here. Maybe these should be less technical and more informative ie "Create  components that automate network synchronization for your networking application"?

AMZN-puvvadar replied:
That's really feedback for @sptramer and @willihay . I was just adding a link here though I'll soon be migrating it to AzNetworking.

lmbr-pip commented on content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autopackets.md:
```diff
@@ -0,0 +1,147 @@
+linktitle: AutoPackets
+---
+
+*AutoPackets* provide a convenient way to define network packets used to communicate data between endpoints in a Multiplayer session. Using the [AzAutoGen](/docs/user-guide/engine/autogen) system, autopacket files found inside of your project are processed during builds to create C++ classes and handlers for packets with defined payloads.
```
To me its not clear what an auto-packet is here? Are we defining a protocol for communication? Are they are collecting of attributes for a packet?

Could you provide an example(s) of a 'typical' autopacket here ie a Connect packet that contains the properties expected during connection events such as a protocol version and ticket id?

AMZN-puvvadar replied:
There's an example detailing that further down in the doc.

lmbr-pip reviewed (commented):
LG, mostly.

Feel some of this text jumps straight to technical details without setting enough context for folks new to O3DE networking. I would try and define what an AutoComponent and AutoPacket are and could be used for in a general way, before getting into the technical specifics of O3DE

AMZN-Gene commented on content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autocomponents.md:
```diff
@@ -1,16 +1,16 @@
 ---
 
-*Autocomponents* provide a convenient way to define states of a multiplayer component relevant to network synchronization. Using the [AzAutoGen](/docs/user-guide/engine/autogen) system, autocomponent files found inside of your project are processed during builds to create C++ classes for components and controllers that provide network replication and remote function calls. Autocomponents also tak ...
+*AutoComponents* provide a convenient way to define states of a multiplayer component relevant to network synchronization. Using the [AzAutoGen](/docs/user-guide/engine/autogen) system, autocomponent files found inside of your project are processed during builds to create C++ classes for components and controllers that provide network replication and remote function calls. AutoComponents also tak ...
```
I think we should spell autocomponent as auto-component.  pinging @sptramer

AMZN-puvvadar replied:
I went with our in code naming but I'm open here. I just felt like Autocomponents and Autopackets didn't necessarily read well.

AMZN-Gene commented on content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autopackets.md:
```diff
@@ -0,0 +1,147 @@
+---
+title: Multiplayer AutoPackets
```
Similar to auto-components I think AutoPacket = Auto-Packet or Auto-packet?

sptramer commented on content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autopackets.md:
```diff
@@ -0,0 +1,147 @@
+
+### Packet attributes
+
+The `Packet` tag defines the name, description and handshake relevance of the packet
+
+| Property | Description | Type |
+|---|---|---|---|
+| Name | The class name of the generated autopacket. | `string`: Must be a valid C++ class name. |
+| Desc | The description of the generated autopacket. | `string`: A user friendly string description of the packet. |
+| HandshakePacket | If the packet is part of a connection handshake. | `string`: Only packets with HandshakePacket as true can be processed while the user implemented IsHandshakeComplete function returns false. |
```
```suggestion
The `Packet` tag defines a new packet.

| Property | Description | Type |
|---|---|---|---|
| Name | The class name of the generated autopacket. | A valid C++ class identifier. |
| Desc | The description of the generated autopacket. | `string` |
| HandshakePacket | If the packet is part of a connection handshake. | `bool` |

{{< note >}}
Packets with `HandshakePacket ` set as true will only be processed if the user-implemented `IsHandshakeComplete` function for the packet returns `false`.
{{< /note >}}
```

sptramer commented on content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autopackets.md:
```diff
@@ -0,0 +1,147 @@
+
+### Member attributes
+
+The `Member` tag defines a member on the packet. This is the primary mechanism for defining a packet's payload.
+
+| Property | Description | Type |
+|---|---|---|---|
+| Name | The name of the packet member. | `string`: Must be a valid C++ variable name. |
+| Type | The type of the packet member. | `string`: Must be a valid C++ type. |
+| Init | The initial value of the packet member. | `<Type>`: Must be a valid value of `<Type>`. |
+| Container | If specified, the member represents a container of `<Type>`. | `Vector`, `Array`|
+| | **Vector**: A resizeable vector. |
+| | **Array**: A fixed size array. | |
+| Count | If Container is specified, the size of the container. | `uint` |
```
```suggestion
The `Member` tag defines data on the packet. This is the primary mechanism for defining a packet's payload.

| Property | Description | Type |
|---|---|---|---|
| Name | The name of the packet member. | A valid C++ variable name. |
| Type | The type of the packet member. | A valid C++ type. |
| Init | The initial value of the packet member. | `<Type>` |
| Container | If provided, the member is a container of `<Type>`. | `Vector`, `Array`|
| | **Vector**: A resizeable vector. |
| | **Array**: A fixed size array. | |
| Count | If Container is provided, the size of the container. | `uint` |
```

sptramer reviewed (changes requested):
* Breaks the link on `gems/reference/_index.md` to the Multiplayer gem

Some suggestions to update the descriptions on the auto-packets. I think I got the technical details correct based on a quick source browse but might have missed something.

AMZN-puvvadar commented on content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autocomponents.md:
```diff
@@ -1,16 +1,16 @@
 ## Autocomponent file structure
 
-Autocomponents are defined in XML files, placed in the `Code\Source\Autogen` directory of the Multiplayer Gem.
+AutoComponents are defined in XML files, placed in the `Code\Source\Autogen` directory of the Multiplayer Gem.
```
Ah yeah, they can be anywhere using AzNetworking technically. I should probably restructure this.

AMZN-puvvadar commented on content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autopackets.md:
```diff
@@ -0,0 +1,147 @@
+
+In order to enable autopacket builds for your project, follow the instructions in [Multiplayer Project Configuration](./configuration).
+
+## AutoPacket file structure
```
Not to my knowledge, unless there was one for AutoComponents that's not apparent to me?

AMZN-puvvadar commented on content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autopackets.md:
```diff
@@ -0,0 +1,147 @@
+
+## AutoPacket file structure
+
+AutoPackets are defined in XML files, placed in the `Code\Source\Autogen` directory of the Multiplayer Gem.
```
They can be anywhere making use of AzNetworking including AzNetworking itself, I'll do some restructuring here.

AMZN-puvvadar commented:
Feedback is largely addressed. I handled the naming format for auto-component and auto-packet and moved the auto-packet doc to networking where it should have been in the first place. I also added links and took much of @sptramer 's formatting and content suggestions.

lmbr-pip commented on content/docs/user-guide/gems/reference/_index.md:
```diff
@@ -123,7 +123,6 @@ toc: true
 | [Certificate Manager](./network/certificate-manager) | The Certificate Manager Gem provides access to authentication files for secure game connections from Amazon S3, files on disk, and other 3rd party sources. |
 | [Http Requestor](./network/http-requestor) | The HTTP Requestor Gem provides functionality to make asynchronous HTTP/HTTPS requests and return data through a user-provided call back function. |
 | [Metastream](./network/twitch/metastream) | The Metastream Gem provides functionality for an HTTP server that allows broadcasters to customize game streams with overlays of statistics and event data from a game session. |
-| [Multiplayer](./network/multiplayer) | The Multiplayer Gem provides support for synchronizing O3DE components across a network in a fault-tolerant way, allowing for the development of multiplayer games. |
```
Did you mean to remove this reference? Don't we want a reference to the Multiplayer Gem here?

AMZN-puvvadar replied:
If you look up a bit, it's in the Multiplayer section right above Networking. It seemed redundant to have it in both.

lmbr-pip commented on content/docs/user-guide/networking/autopackets.md:
```diff
@@ -0,0 +1,151 @@
+
+### PacketGroup attributes
+
+The top-level element of an auto-packet definition is a `PacketGroup`. Packet groups associate different types of individual packet definitions together. Packets defined as part of a packet group are placed into a namespace.
```
nit: "the same namespace" or " the packet group's namespace"? to be clearer.

lmbr-pip commented on content/docs/user-guide/networking/autopackets.md:
```diff
@@ -0,0 +1,151 @@
+
+| Property | Description | Type |
+|---|---|---|---|
+| Name | The class name of the generated autopacket. | A valid C++ class identifier. |
```
Do any of these fields have size/limits on the type?

AMZN-puvvadar replied:
Nope, should just be the limitations of the type itself.

lmbr-pip commented on content/docs/user-guide/networking/autopackets.md:
```diff
@@ -0,0 +1,151 @@
+
+| Property | Description | Type |
+|---|---|---|---|
+| Name | The name of the packet member. | A valid C++ variable name. |
```
Do any of these fields have size/limits on the type?

AMZN-puvvadar replied:
Nope, these aren't specialized types so there's nothing explicit. Granted larger counts on containers will yield larger packets.

lmbr-pip commented on content/docs/user-guide/networking/autopackets.md:
````diff
@@ -0,0 +1,151 @@
+```C++
+    //! IConnectionListener interface
+    //! @{
+    AzNetworking::ConnectResult ValidateConnect(const AzNetworking::IpAddress& remoteAddress, const AzNetworking::IPacketHeader& packetHeader, AzNetworking::ISerializer& serializer) override;
````
nit: Consider new lines between methods to make this code block more readable

lmbr-pip reviewed (commented):
LG.

Have some minor comments.
